File: 13-roman-to-integer/roman-to-integer.php
class Solution {

    /**
     * @param String $s
     * @return Integer
     */
    function romanToInt($s)
{
    $track = array(
        "I" => 1,
        "V" => 5,
        "X" => 10,
        "L" => 50,
        "C" => 100,
        "D" => 500,
        "M" => 1000
    );
    $keys = array_keys($track);
    $n = strlen($s);
    $ans = 0;
    for ($i = 0; $i < $n; ++$i) {
        if ($i + 1 < $n and $s[$i] == 'I' and ($s[$i + 1] == "V" or $s[$i + 1] == "X")) {
            $ans += $track[$s[$i + 1]] - $track['I'];
            ++$i;
        } else if ($i + 1 < $n and $s[$i] == 'X' and ($s[$i + 1] == "L" or $s[$i + 1] == "C")) {
            $ans += $track[$s[$i + 1]] - $track['X'];
            ++$i;
        } else if ($i + 1 < $n and $s[$i] == 'C' and ($s[$i + 1] == "D" or $s[$i + 1] == "M")) {
            $ans += $track[$s[$i + 1]] - $track['C'];
            ++$i;
        } else {
            $ans += $track[$s[$i]];
        }
    }
    return $ans;
}
}

File: 13-roman-to-integer/roman-to-integer.java
class Solution {

    public int romanToInt(String s) {
        Map<Character, Integer> track = new HashMap<>();
        track.put('I', 1); track.put('V', 5);
        track.put('X', 10); track.put('L', 50);
        track.put('C', 100); track.put('D', 500);
        track.put('M', 1000);

        int n = s.length();
        int ans = 0;

        for (int i = 0; i < n; ++i) {
            if (i + 1 < n && s.charAt(i) == 'I' && (s.charAt(i + 1) == 'V' || s.charAt(i + 1) == 'X')) {
                ans += track.get(s.charAt(i + 1)) - track.get('I');
                ++i;
            } else if (i + 1 < n && s.charAt(i) == 'X' && (s.charAt(i + 1) == 'L' || s.charAt(i + 1) == 'C')) {
                ans += track.get(s.charAt(i + 1)) - track.get('X');
                ++i;
            } else if (i + 1 < n && s.charAt(i) == 'C' && (s.charAt(i + 1) == 'D' || s.charAt(i + 1) == 'M')) {
                ans += track.get(s.charAt(i + 1)) - track.get('C');
                ++i;
            } else {
                ans += track.get(s.charAt(i));
            }
        }

        return ans;
    }
}